### PR TITLE
MBL-1022: Add blocking mutation to iOS client

### PR DESCRIPTION
**Pull request by amy-at-kickstarter**
# 📲 What

Add `BlockUser` mutation to the app.

# 🤔 Why

So we can implement blocking in the UI!

### CODE REVIEW

File: KsApi/Service.swift
Comment (amy-at-kickstarter) [inline KsApi/Service.swift:799]: So `BlockUser` _can_ return two fields: `success` and `currentUser { blockedUsers } `. However, based on my current understanding of the client designs, we don't need to read either of these fields: we'll fetch `isBlocked` directly on other `User` objects.

Is that correct? Based on that assumption I figured I'd take the simplest route out and return an `EmptyResponseEnvelope`.

File: KsApi/Service.swift
Comment (amy-at-kickstarter) [inline KsApi/Service.swift:797]: n.B. `GraphAPI.BlockUserMutation` wants us to pass it a `GraphQLID` type, which can also be a `String`. It's not obvious to me if we'll need to base64-encode the value like a proper GraphQL identifier, or if it will be happy taking a plain, string-based user ID like `"12345"`.

Comment (scottkicks) [reply KsApi/Service.swift:797]: Let's follow the same "adapter" pattern that we're using elsewhere here. We can create an extension on `GraphAPI.BlockUserInput` for creating the input type. Extracting this allows us to test it more easily. See `GraphAPI.WatchProjectInput.from(_ input:...` for an example

File: KsApi/Service.swift
Comment (scottkicks) [reply KsApi/Service.swift:799]: Yep exactly. We only need to know if it was successful right now.

File: KsApi/ServiceType.swift
Comment (scottkicks) [inline KsApi/ServiceType.swift:384]: So, this will be `func blockUser(input: BlockUserInput) -> SignalProducer<EmptyResponseEnvelope, ErrorEnvelope>`

**Review by scottkicks — COMMENTED** [review]
Nice first run through! There's just a couple of things we need to add to stick to the existing pattern and keep things well tested. I added inline comments and you can use the linked PR below as a guide. 

Let's create a `BlockUserInput` type to inject into our Service method as well as a `.from(_ input...)` helper extension to make it easier on ourselves to bundle up the expected properties once we implement this. Check out [this PR](https://github.com/kickstarter/ios-oss/pull/1792) out as a reference

Oh and we can add some tests around the input and extension as well 😄

File: KsApi/GraphAPI.swift
Comment (amy-at-kickstarter) [reply KsApi/GraphAPI.swift:6844]: One point worth noting - right now we only update the schema when we _need_ to update it on the client. But we should really be updating it any time changes to the web schema are pushed out - otherwise, we risk getting out of sync, and (say) using a feature that no longer actually exists on the web.

Obviously I'd hope that someone working on the web side would point this out to us before merging that change! But nice to enforce the same thing through code, not just socially.

File: KsApi/models/graphql/adapters/GraphAPI.BlockUserInput+BlockUserInput.swift
Comment (amy-at-kickstarter) [inline KsApi/models/graphql/adapters/GraphAPI.BlockUserInput+BlockUserInput.swift:1]: For what it's worth - I'm happy to follow the pattern for now, but considering how simple this input is, I'm not entirely sure it's worth having this extra boilerplate for the adapter + adapter tests. Just something to consider in the back of our minds as we move forward with other kinds of API refactoring in the future.

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: KsApi/mutations/inputs/BlockUserInput.swift
Comment (amy-at-kickstarter) [inline KsApi/mutations/inputs/BlockUserInput.swift:1]: @scottkicks Can you explain a little more about this input adapter pattern? Happy to follow it for now, but at least for a very teeny input like this, it doesn't seem obviously beneficial. I'm curious to hear more about why we've been using this pattern instead of using the `GraphAPI.XYZInput` objects directly.

**Comment by scottkicks** [comment]
@amy-at-kickstarter Until we can say this this current mutation is working/testable we might want to hold off on merging this. In case Santiago has to make further changes. ref: https://kickstarter.slack.com/archives/C05HX9GS0CD/p1701219627599909?thread_ts=1701184665.560489&cid=C05HX9GS0CD

**Review by scottkicks — APPROVED** [review]

**Pull request merged**
